### DISCUSSION TITLE
Connection between vitrivr-ng and cineast is lost, when running vitrivr-ng on an ssh apache server.

### DISCUSSION THREAD

**Discussion by RibinMathewChalumattu** [ppanopticon/vitrivr-ng#64] [Q&A] [answered]
Dear devs,

I wanted to deploy vitrivr-ng on a local apache2 Ubuntu server. Following the instruction on the wiki, I could deploy the web interface and the connection to cineast also worked. But after enabling ssh on the apache server, vitrivr-ng cannot establish a connection to cineast. Do I need to modify anything in the config files to establish a connection, or is this a configuration issue  in the apache server? 

Best,

Ribin

**Comment by lucaro** [comment]
Ssh is an independent process and should have nothing to do with your connection problem. Make sure that the cineast API port is reachable from outside your server and that vitrivr-ng uses the server address as a destination, rather than your local address.

**Comment by RibinMathewChalumattu** [comment, marked as answer]
Cineast is reachable when I run vitrivr-ng  in the apache server under http://server_ip, but not when I run it with https://server_ip.

**Reply by lucaro** [reply]
Ah, so you mean SSL/TLS rather than SSH. In that case, it probably is your browser refusing to connect to an unsecured websocket if the rest of the page is served via HTTPS. You can verify this by looking at the error messages in your browser's developer console. If you want to use an encrypted connection between cineast and vitrivr-ng, you can either configure cineast to offer the API via HTTPS and WSS directly or configure your apache as a proxy for the API and websocket connection.

**Reply by RibinMathewChalumattu** [reply]
Thanks, indeed the problem was the browser refusing to connect. Do you have instructions on how to configure cineast to offer the API via HTTPS? I have seen that in the _cineast.json_ there is a property _enableWebsocketSecure_ under the _api_ property. How do I setup a ssl certificate with cineast?

**Reply by lucaro** [reply]
To make HTTPS work, you need to generate a java key store file with your certificates. Then simply specify the path to the keystore file in the config, using the `keystore` parameter. The parameter for the password of the key store is `keystorePassword`, see https://github.com/vitrivr/cineast/blob/master/cineast-runtime/src/main/java/org/vitrivr/cineast/standalone/config/APIConfig.java for further reference.

**Reply by RibinMathewChalumattu** [reply]
For the proxy-based approach, I have read that the "[ProxyPass](https://httpd.apache.org/docs/current/mod/mod_proxy.html)" can be used in the virtual host configuration file to connect to an external WebSocket.  I have tried the following, but without success: " _ProxyPass "/ws" "ws://129.132.62.166:4567/api/v1/websocket"_
_ProxyPassReverse "/ws" "ws://129.132.62.166:4567/api/v1/websocket"_.

The "ws://129..." corresponds to the cineast WebSocket address, as shown in the vitrivr-web interface. I'm not what the correct parameter is for the "/ws" part. Do you have an idea? Is this the right approach to get the proxy to work?

**Reply by lucaro** [reply]
We can't really offer support for configuring apache here, but if you are running cineast on the same machine as your apache instance, it should probably rather be something like ProxyPass "/ws" "ws://127.0.0.1:4567/api/v1/websocket". Also, this alone will not solve your actual problem, since you'd also need to set up a proxy which translates the 'ws' of cineast into 'wss' on the apache side. Taking things one step at a time, I'd recommend to first disable SSL on the apache and try to get the proxy working without encryption. Once that's done, you know a configuration that works and does some of what you want. Then you can adapt it to do the same thing also with wss.

**Reply by RibinMathewChalumattu** [reply]
Thanks for the replies. In the end I went with the HTTPS and keystore route and the connection between vitrivr-ng and cineast could be established.
